artichoke-backend/src/extn/core/string/scan.rs

```diff
@@ -11,94 +12,6 @@ use crate::types::Ruby;
 use crate::value::{Value, ValueLike};
 use crate::Artichoke;
 
-fn literal_scan_count(string: &[u8], pattern: &[u8]) -> (usize, usize) {
-    if pattern.is_empty() {
-        (string.len() + 1, string.len())
-    } else if pattern.len() > string.len() {
-        (0, 0)
-    } else if pattern == string {
-        (1, 0)
-    } else {
-        match pattern.len() {
-            0 => unreachable!("handled above"),
-            1 => {
-                let mut matches = 0;
-                let mut last_pos = 0;
-                let byte0 = pattern[0];
-                for pos in memchr::memchr_iter(byte0, string) {
-                    matches += 1;
-                    last_pos = pos;
-                }
-                (matches, last_pos)
-            }
-            _ => {
-                let mut matches = 0;
-                let mut last_pos = 0;
-                let byte0 = pattern[0];
-                let rest = &pattern[1..];
-                let strlen = string.len();
-                let patlen = pattern.len();
-                let mut start = 0;
-                while let Some(pos) = memchr::memchr(byte0, &string[start..]) {
-                    last_pos = pos;
-                    let idx = start + pos;
-                    start = idx + 1;
-                    if idx + patlen <= strlen {
-                        if &string[idx + 1..idx + patlen] == rest {
-                            matches += 1;
-                            start = idx + patlen;
-                        }
-                    } else {
-                        break;
-                    }
-                }
```

Comment:
Just as a note for future reference: the other issue with this implementation of substring search is that it will degrade pretty rapidly in certain pathological cases. e.g., with a needle of `aaaaaaaab` on a haystack of `aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...`. You'll basically wind up comparing the needle at every position of the input. This sort of behavior is endemic to approaches which use memchr acceleration, and serves as a cautionary tale for benchmarking. It's very easy to do well on certain cases while performing _much worse_ in other cases. This is why benchmarking substring search is so difficult. Your benchmarks need a good sampling of both needles and haystacks.

bstr's implementation has protections against the sub-optimal behavior, mostly by keeping track of how effective the memchr acceleration is. If it turns out that memchr is skipping only a small number of bytes on average, then the search will dynamically disable the use of memchr and fallback to the standard Two-Way algorithm.